chain-signatures/node/src/indexer_eth.rs

```diff
@@ -258,57 +260,165 @@ pub async fn run(
+                        tracing::info!("Latest eth sign request block number: {latest_block_number}");
+                        crate::metrics::LATEST_BLOCK_NUMBER
+                            .with_label_values(&[Chain::Ethereum.as_str(),node_near_account_id.as_str()])
+                            .set(latest_block_number as i64);
+                    }
+                    _ = heartbeat_interval.tick() => {
+                        if latest_sign_request_time.elapsed() > heartbeat_interval.period() {
+                            tracing::warn!("No sign request received in the last 180 seconds, unsubscribing...");
+                            break;
                         }
                     }
-                    Err(err) => tracing::warn!("Failed to subscribe to logs: {:?}", err),
                 }
             }
-            Err(err) => tracing::error!("Failed to connect to Ethereum WebSocket: {:?}", err),
+            let Ok(unsubscribed) = filtered_logs_sub.unsubscribe().await.inspect_err(|err| {
+                tracing::warn!(
+                    "Failed to unsubscribe from logs: {:?}, will reconnect to websocket",
+                    err
+                );
+            }) else {
+                break;
+            };
+
+            if !unsubscribed {
+                tracing::warn!("Unsubscribe from logs return false, will reconnect to websocket");
+                break;
+            }
+        }
+    }
+}
+
+fn process_filtered_log(
+    log: web3::types::Log,
+    sign_tx: mpsc::Sender<IndexedSignRequest>,
+    node_near_account_id: AccountId,
+) -> anyhow::Result<()> {
+    tracing::info!("Received new Ethereum sign request: {:?}", log);
+    let sign_request = sign_request_from_filtered_log(log)?;
+    let sign_tx = sign_tx.clone();
+    tokio::spawn(async move {
+        if let Err(err) = sign_tx.send(sign_request).await {
+            tracing::error!(?err, "Failed to send ETH sign request into queue");
+        } else {
+            crate::metrics::NUM_SIGN_REQUESTS
+                .with_label_values(&[Chain::Ethereum.as_str(), node_near_account_id.as_str()])
+                .inc();
         }
+    });
+    Ok(())
+}
+
+async fn catchup(
+    start_block: u64,
+    end_block: u64,
+    ws: Web3<WebSocket>,
+    sign_tx: mpsc::Sender<IndexedSignRequest>,
+    filter_builder: FilterBuilder,
+    node_near_account_id: AccountId,
+) -> anyhow::Result<()> {
+    tracing::info!("Catching up from block {start_block} to block {end_block}");
+    let filter = filter_builder
+        .from_block(BlockNumber::Number(start_block.into()))
+        .to_block(BlockNumber::Number(end_block.into()))
+        .build();
 
-        tracing::warn!("Ethereum WebSocket disconnected, reconnecting in 2 seconds...");
-        tokio::time::sleep(Duration::from_secs(2)).await;
+    let logs = ws.eth().logs(filter).await?;
+    for log in logs {
+        process_filtered_log(log, sign_tx.clone(), node_near_account_id.clone())?;
     }
```

Comment:
on failing to catchup, would it be better to just warn that we weren't able to process this specific log, because the moment we early return with `?`, the next entries for logs aren't handled. Will they get handled in the next call to `catchup`?